Source/RapyutaSimulationPlugins/Private/Sensors/ROS2Camera.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020-2021 Rapyuta Robotics Co., Ltd.
+
+
+#include "Sensors/ROS2Camera.h"
+
+AROS2Camera::AROS2Camera()
+{
+}
+
+void AROS2Camera::BeginPlay()
+{
+	Super::BeginPlay();	
+    Init();
+}
+
+void AROS2Camera::Init()
+{
+    Super::Init();
+
+    // component initialization
+    SceneCaptureComponent = NewObject<USceneCaptureComponent2D>(this, USceneCaptureComponent2D::StaticClass());
+    // SceneCaptureComponent->RegisterComponent(); // not required?
+    RenderTarget = NewObject<UTextureRenderTarget2D>(this, UTextureRenderTarget2D::StaticClass());
+
+    // Initialize capture and texture component
+    SceneCaptureComponent->SetWorldTransform(this->GetTransform());
+    SceneCaptureComponent->FOVAngle = FOV;
+    SceneCaptureComponent->OrthoWidth = OrthWidth;
+    
+    RenderTarget->InitCustomFormat(Width, Height, EPixelFormat::PF_B8G8R8A8, true);
+    SceneCaptureComponent->TextureTarget = RenderTarget;
+
+    // Initialize image data
+    Data.width = Width;
+    Data.height = Height;
+    Data.encoding = Encoding;
+    Data.step = Width * 3; // todo should be variable based on encoding
+    Data.data.AddUninitialized(Width * Height * 3);
+
+    // ROS Image topic publisher
+    Publisher =  NewObject<UROS2Publisher>(this, UROS2Publisher::StaticClass());
+    Publisher->RegisterComponent();
+    Publisher->TopicName = TopicName;
+    Publisher->PublicationFrequencyHz = PublishFreq;
+    Publisher->MsgClass = UROS2ImageMsg::StaticClass();
+    
+    Publisher->UpdateDelegate.BindDynamic(this, &AROS2Camera::MessageUpdate);
+    AddPublisher(Publisher);
+    Publisher->Init(UROS2QoS::KeepLast);
+
+}
+
+void AROS2Camera::MessageUpdate(UROS2GenericMsg *TopicMessage)
+{
+    UROS2ImageMsg *Message = Cast<UROS2ImageMsg>(TopicMessage);
+    Message->SetMsg(GetData());
+}
+
+FROSImage AROS2Camera::GetData()
+{
+    SceneCaptureComponent->CaptureScene();
+
+    FTextureRenderTarget2DResource* RenderTargetResource;
+    RenderTargetResource = (FTextureRenderTarget2DResource*)RenderTarget->GameThread_GetRenderTargetResource();
+    if (RenderTargetResource) {
+        TArray<FColor> buffer;
+        RenderTargetResource->ReadPixels(buffer);

```

Comment:
Just a note, this would be blocking if called in high frequency. A non-blocking version could be referenced [here](https://github.com/TimmHess/UnrealImageCapture).
Some discussions on a related feature being implemented in AirSim:
https://github.com/microsoft/AirSim/pull/3018
https://github.com/Microsoft/AirSim/issues/1818
https://github.com/microsoft/AirSim/blob/master/Examples/DataCollection/StereoImageGenerator.hpp

Other related:
https://github.com/ayumax/WindowCapture2D
https://github.com/rnixik/StereoPano
https://blog.jamie.holdings/2019/09/08/taking-screenshots-in-game-in-unreal-engine-4-23/